### PR TITLE
Add support for stm32f107

### PR DESCRIPTION
**Pull request by franeklubi**
This is me reviving an old pull request: #23 

My changes include:
* synchronizing with changes on master
* updating the dependencies to the newest minor versions and fixing all breakages
* fixing examples

### CODE REVIEW

**Review by rfael — COMMENTED** [review]
Great work! Despite two small things, everything looks good for me.

File: .cargo/config [hunk 1]
```diff
@@ -4,9 +4,15 @@ rustflags = [
...
+]
+
 [build]
 # Pick ONE of these compilation targets
 # target = "thumbv6m-none-eabi"    # Cortex-M0 and Cortex-M0+
-# target = "thumbv7m-none-eabi"    # Cortex-M3
+target = "thumbv7m-none-eabi"    # Cortex-M3
 # target = "thumbv7em-none-eabi"   # Cortex-M4 and Cortex-M7 (no FPU)
-target = "thumbv7em-none-eabihf" # Cortex-M4F and Cortex-M7F (with FPU)
+# target = "thumbv7em-none-eabihf" # Cortex-M4F and Cortex-M7F (with FPU)
```
Comment (mateuszkj) [inline .cargo/config:18]: I think default target should be as it was. (Cortex-M4F)

File: Cargo.toml [hunk 2]
```diff
@@ -4,7 +4,7 @@ name = "stm32-eth"
 description = "Embedded Rust Ethernet driver for the STM32 MCU series"
 license = "Apache-2.0"
 authors = ["Astro <astro@spaceboyz.net>"]
-version = "0.2.0"
+version = "0.3.0"
```
Comment (mateuszkj) [inline Cargo.toml:7]: revert it to version 0.2.0 as in master branch.

File: Cargo.toml [hunk 3]
```diff
@@ -36,6 +37,8 @@ device-selected = []
 fence = []
 smi = []
 
+stm32f107 = ["stm32f1xx-hal/stm32f107", "device-selected"]
+
```
Comment (mateuszkj) [inline Cargo.toml:41]: remove empty line

File: examples/arp-smoltcp.rs [hunk 4]
```diff
@@ -12,24 +12,23 @@ extern crate panic_itm;
...
+use fugit::RateExtU32;
+use smoltcp::wire::{
+    ArpOperation, ArpPacket, ArpRepr, EthernetAddress, EthernetFrame, EthernetProtocol,
+    EthernetRepr, Ipv4Address,
+};
 use stm32_eth::{
     hal::gpio::{GpioExt, Speed},
     hal::rcc::RccExt,
-    hal::time::U32Ext,
+    // hal::time::U32Ext,
```
Comment (mateuszkj) [inline examples/arp-smoltcp.rs:28]: if not used, remove it

File: memory.x [hunk 5]
```diff
@@ -6,6 +8,13 @@ MEMORY
...
 _stack_start = ORIGIN(RAM) + LENGTH(RAM);
 _heap_size = LENGTH(RAM) - 4K;
+*/
+
+/* STM32F107 */
+MEMORY
+{
+  FLASH : ORIGIN = 0x08000000, LENGTH = 128K
+  RAM : ORIGIN = 0x20000000, LENGTH = 32K
+}
```
Comment (mateuszkj) [inline memory.x:20]: I think is should be reverted default `memory.x` to define STM32F429 and commented for STM32F107. Or there should be another file `memory_stm32f107.x`

File: src/lib.rs [hunk 6]
```diff
@@ -163,11 +173,16 @@ impl<'rx, 'tx> Eth<'rx, 'tx> {
...
         });
+
         // Flow Control Register
         self.eth_mac.macfcr.modify(|_, w| {
             // Pause time
-            w.pt().bits(0x100)
+            #[allow(unused_unsafe)]
+            unsafe {
+                w.pt().bits(0x100)
+            }
```
Comment (mateuszkj) [inline src/lib.rs:183]: is `allow(unused_unsafe)` necessary ? is unsafe necessary?

File: src/rx.rs [hunk 7]
```diff
@@ -239,6 +234,7 @@ impl<'a> RxRing<'a> {
         // Register RxDescriptor
         eth_dma
             .dmardlar
+            // Note: unsafe block required for `stm32f107`.
             .write(|w| unsafe { w.srl().bits(ring_ptr as u32) });
```
Comment (mateuszkj) [inline src/rx.rs:238]: is there clippy warning for others CPUs ? maybe it should be added `allow(unused_unsafe)`

File: Cargo.toml (hunk 3 above)
Comment (franeklubi) [reply Cargo.toml:41]: I think it's better to divide between `stm32f1xx`, `stm32f4xx`, and `stm32f7xx` as it is now

File: Embed.toml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+[default.probe]
```
[the remark this answers is not in the document]
Comment (franeklubi) [reply Embed.toml:1]: This file is used by `probe-rs/cargo-embed` - a tool for working with embedded targets

It must've showed up in some previous commit and I agree, it shouldn't be here

File: src/lib.rs (hunk 6 above)
Comment (franeklubi) [reply src/lib.rs:183]: It probably was before, but it's not needed any more.

File: src/rx.rs (hunk 7 above)
Comment (franeklubi) [reply src/rx.rs:238]: No warnings for any other CPU - I'm removing the note

**Comment by franeklubi** [comment]
Thank you for the review @rfael and @mateuszkj 

I've implemented the changes I marked with `👍` emoji

**Comment by korken89** [comment]
Pinging again @thalesfragoso / @adamgreig

File: src/setup.rs [hunk 9]
```diff
@@ -93,6 +93,53 @@ pub(crate) fn setup() {
         rcc.ahb1rstr.modify(|_, w| w.ethmacrst().set_bit());
         rcc.ahb1rstr.modify(|_, w| w.ethmacrst().clear_bit());
     });
+
+    #[cfg(feature = "stm32f1xx-hal")]
+    cortex_m::interrupt::free(|_| unsafe {
```
Comment (thalesfragoso) [inline src/setup.rs:98]: I would prefer this to be done with bit banding, but we can leave that for another PR.

**Review by thalesfragoso — COMMENTED** [review]
Thank you and all the others that reviewed this.

I just left a few questions.

File: src/setup.rs [hunk 10]
```diff
@@ -204,3 +255,49 @@ impl_pins!(
...
+    unsafe impl RmiiRxD0 for PC4 {}
+    unsafe impl RmiiRxD0 for PD9 {}
+    unsafe impl RmiiRxD1 for PC5 {}
+    unsafe impl RmiiRxD1 for PD10 {}
+
+    macro_rules! impl_alt_very_high_speed {
+        ($($PIN:ident),*) => {
+            $(
+                impl AlternateVeryHighSpeed for $PIN {
+                    fn into_af11_very_high_speed(self) {}
```
[the remark this answers is not in the document]
Comment (korken89) [reply src/setup.rs:296]: This is guaranteed by `Table 209. Ethernet pin configuration`

[the remark this answers is not in the document]
Comment (oyren) [reply src/setup.rs:296]: The initial switch to the git dependency came from me and was a mistake which no one cateched. 
At this state it wasn't required.

I can't say for sure what the current status is but i think it is not necessary and the current published of the stm32f7-hal is sufficient.

(This is systec-ms private account)

File: Cargo.toml [hunk 11]
```diff
@@ -15,13 +15,14 @@ travis-ci = { repository = "astro/stm32-eth", branch = "master" }
...
 [package.metadata.docs.rs]
-features = [ "smi", "smoltcp-phy", "stm32f429", "smoltcp/socket-tcp" ]
+features = ["smi", "smoltcp-phy", "stm32f429", "smoltcp/socket-tcp"]
 
 [dependencies]
 volatile-register = "0.2"
 aligned = "0.4"
-stm32f7xx-hal = {git = "https://github.com/stm32-rs/stm32f7xx-hal", rev = "d632a5c", optional = true}
-stm32f4xx-hal = {version = "0.10", optional = true}
+stm32f7xx-hal = { git = "https://github.com/stm32-rs/stm32f7xx-hal", rev = "7099b07", optional = true }
```
[the remark this answers is not in the document]
Comment (oyren) [reply Cargo.toml:23]: The initial switch to the git dependency came from me and was a mistake which no one cateched.
At this state it wasn't required.

I can't say for sure what the current status is but i think it is not necessary and the current published of the stm32f7-hal is sufficient.

(This is systec-ms private account)
